### PR TITLE
added support for Redshift Clusters

### LINKED ISSUE #333: AWS Redshift Cluster support

### PR DESCRIPTION
**Pull request by Nirbhay1997**
Signed-off-by: nirbhay <nkumar@chef.io>

### Description

Added Support for Aws Redshift Cluster
### Issues Resolved

#333 

### Check List
Please fill box or appropriate ([x]) or mark N/A.
- [x] New functionality includes integration tests/controls
- [x] New Terraform resources
- [x] Documentation provided or updated for resources 
- [x] All Integration Tests pass
- [x] All Unit Tests pass
- [x] `rake lint` passes
- [x] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>

### CODE REVIEW

File: docs/resources/aws_redshift_cluster.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    end
+
+#### Parameters
+
+##### cluster\_identifier _(required)_
+
+This resource accepts a single parameter, the user-supplied cluster identifier. This parameter isn't case-sensitive.
+This can be passed either as a string or as a `cluster_identifier: 'value'` key-value entry in a hash.
+
+See also the [AWS documentation on Redshift cluster](https://docs.aws.amazon.com/AmazonRedshift/latest/AuroraUserGuide/Aurora.Overview.html).
```
Comment (IanMadd) [inline docs/resources/aws_redshift_cluster.md:30]: This link is bad. Is https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/Aurora.Overview.html the correct link?

File: libraries/aws_redshift_cluster.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    describe aws_redshift_cluster(cluster_identifier: 'test-cluster-id') do
+      it { should exist }
+    end
+  "
+  def initialize(opts = {})
+    opts = { cluster_identifier: opts } if opts.is_a?(String)
+    super(opts)
+    validate_parameters(required: [:cluster_identifier])
+
+    raise ArgumentError, "#{@__resource_name__}: cluster_identifier must start with a letter followed by up to 62 letters/numbers/hyphens." if opts[:cluster_identifier] !~ /^[a-z]{1}[0-9a-z\-]{0,62}$/
```
Comment (soumyo13) [inline libraries/aws_redshift_cluster.rb:19]: Write this also in the documentation for both singular and plural resources.

File: libraries/aws_redshift_clusters.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  example '
+    describe aws_redshift_clusters do
+      it { should exist }
+    end
+  '
+
+  attr_reader :table
+
+  FilterTable.create
+             .register_column(:cluster_identifier, field: :cluster_identifier)
```
Comment (soumyo13) [inline libraries/aws_redshift_clusters.rb:17]: Change the names to plural names as this is a plural resource.

File: libraries/aws_redshift_clusters.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                           master_username:                  redshift_cluster.master_username,
+                           db_name:                          redshift_cluster.db_name,
+                           cluster_create_time:              redshift_cluster.cluster_create_time,
+                           cluster_subnet_group_name:        redshift_cluster.cluster_subnet_group_name,
+                           vpc_id:                           redshift_cluster.vpc_id,
+                           allow_version_upgrade:            redshift_cluster.allow_version_upgrade,
+                           encrypted:                        redshift_cluster.encrypted,
+                           iam_roles:                        redshift_cluster.iam_roles,
+                           modify_status:                    redshift_cluster.modify_status,
+                           availability_zone:               redshift_cluster.availability_zone }]
```
Comment (soumyo13) [inline libraries/aws_redshift_clusters.rb:59]: ome space missing.

**Review by soumyo13 — CHANGES_REQUESTED** [review]
Update the documentation files in the Readme.

File: libraries/aws_redshift_cluster.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+    catch_aws_errors do
+      @display_name = opts[:cluster_identifier]
+
+      begin
+        resp = @aws.redshift_client.describe_clusters(cluster_identifier: opts[:cluster_identifier])
+        return if resp.clusters.empty?
+        @rds_cluster = resp.clusters[0].to_h
+      rescue Aws::Redshift::Errors::NotAvailableValueError
```
Comment (sathish-progress) [inline libraries/aws_redshift_cluster.rb:28]: please confirm if this inherits from Aws::ServiceErrors and remove this rescue conditions

File: libraries/aws_redshift_cluster.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        return
+      end
+      unless @rds_cluster[:cluster_parameter_groups].to_s.empty?
```
Comment (sathish-progress) [inline libraries/aws_redshift_cluster.rb:31]:
```suggestion
      if  @rds_cluster[:cluster_parameter_groups] && !@rds_cluster[:cluster_parameter_groups].empty?
```

File: libraries/aws_redshift_cluster.rb [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        parameter_groups = @rds_cluster[:cluster_parameter_groups]
+        @rds_cluster[:cluster_parameter_groups] = parameter_groups.map { |x| x.values[0] }
```
Comment (sathish-progress) [inline libraries/aws_redshift_cluster.rb:33]: what are we trying to do here?

File: libraries/aws_redshift_cluster.rb [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+      end
+      create_resource_methods(@rds_cluster)
+    end
+  end
+
+  def has_encrypted?
+    @rds_cluster[:encrypted]
+  end
+
+  alias encrypted? has_encrypted?
```
Comment (sathish-progress) [inline libraries/aws_redshift_cluster.rb:43]: prefer `has_* ` methods only if you plan to implement matchers with arguments. For boolean you should use `be_*` matchers

File: libraries/aws_redshift_clusters.rb [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+             .register_column(:db_name, field: :db_name)
+             .register_column(:modify_status, field: :modify_status)
+             .register_column(:master_username, field: :master_username)
+             .register_column(:cluster_create_time, field: :cluster_create_time)
+             .register_column(:cluster_subnet_group_name, field: :cluster_subnet_group_name)
+             .register_column(:availability_zones, field: :availability_zones)
+             .register_column(:allow_version_upgrade, field: :allow_version_upgrade)
+             .register_column(:encrypted, field: :encrypted)
+             .register_column(:iam_roles, field: :iam_roles)
+             .register_column(:vpc_id, field: :vpc_id)
```
Comment (sathish-progress) [inline libraries/aws_redshift_clusters.rb:30]: column names could be plural

File: test/integration/verify/controls/aws_redshift_cluster.rb [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+  describe aws_redshift_cluster(cluster_identifier: aws_redshift_db_identifier) do
+
+    it { should exist }
+    its ('node_type') { should eq 'dc2.large' }
+    its ('cluster_status') { should eq 'available'  }
+    its ('cluster_availability_status') { should eq 'Available' }
+    its ('master_username') { should eq 'test-cluster' }
+    its ('db_name') { should eq 'dev' }
+    its ('vpc_id') { should eq 'vpc-6d9d7505' }
```
Comment (sathish-progress) [inline test/integration/verify/controls/aws_redshift_cluster.rb:15]: VPC ID could be an attribute

File: test/integration/verify/controls/aws_redshift_clusters.rb [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+  title 'Ensure AWS redshift clusters has the correct properties.'
+
+  describe aws_redshift_clusters do
+    it { should exist }
+    its ('node_type') { should include 'dc2.large' }
+    its ('cluster_status') { should include 'available'  }
+    its ('cluster_availability_status') { should include 'Available' }
+    its ('master_username') { should include 'test-cluster' }
+    its ('db_name') { should include 'dev' }
+    its ('vpc_id') { should include 'vpc-6d9d7505' }
```
Comment (sathish-progress) [inline test/integration/verify/controls/aws_redshift_clusters.rb:13]: same here

**Comment by sathish-progress** [comment]
please also update readme with doc links

File: docs/resources/aws_rds_cluster.md [hunk 12]
```diff
@@ -20,10 +20,10 @@ An `aws_rds_cluster` resource block uses resource parameters to search for an RD
     describe aws_rds_cluster(db_cluster_identifier: 'test-cluster-id') do
       it { should exist }
     end
-    
-#### Parameters
 
-##### db\_cluster\_identifier _(required)_
+### Parameters
```
Comment (sathish-progress) [inline docs/resources/aws_rds_cluster.md:24]: why these changes here ?

**Review by sathish-progress — APPROVED** [review]
LGTM

**Review by soumyo13 — APPROVED** [review]

**Pull request merged**